File: backend_recipe/models/image_generation.py
"""
Image Generation Pydantic Models
=================================
Models for image generation jobs, status, and logs
"""

from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any
from datetime import datetime


# ============================================================================
# Request Models
# ============================================================================

class StartImageGenerationRequest(BaseModel):
    """Request to start batch image generation"""
    image_type: str = Field(
        default="main",
        description="Type of images to generate: 'main', 'steps', or 'all'"
    )
    start_from_recipe_id: Optional[int] = Field(
        None,
        description="Start from specific recipe ID (optional, auto-detects if not provided)"
    )


class StopImageGenerationRequest(BaseModel):
    """Request to stop a running job"""
    job_id: int = Field(..., description="Job ID to stop")


# ============================================================================
# Response Models
# ============================================================================

class ImageGenerationJobModel(BaseModel):
    """Image generation job information"""
    id: int
    status: str
    image_type: str
    started_at: Optional[datetime]
    completed_at: Optional[datetime]
    total_recipes: int
    completed_count: int
    failed_count: int
    skipped_count: int
    current_recipe_id: Optional[int]
    current_recipe_name: Optional[str]
    last_processed_recipe_id: Optional[int]
    start_from_recipe_id: Optional[int]
    should_stop: bool
    error_message: Optional[str]
    error_count: int
    created_at: datetime
    updated_at: datetime


class ImageGenerationLogModel(BaseModel):
    """Image generation log entry"""
    id: int
    job_id: int
    timestamp: datetime
    level: str
    message: str
    recipe_id: Optional[int]
    recipe_name: Optional[str]
    error_details: Optional[Dict[str, Any]]
    metadata: Optional[Dict[str, Any]]


class StartJobResponse(BaseModel):
    """Response when starting a job"""
    success: bool
    message: str
    job_id: Optional[int] = None
    completed: Optional[int] = None
    failed: Optional[int] = None
    skipped: Optional[int] = None


class StopJobResponse(BaseModel):
    """Response when stopping a job"""
    success: bool
    message: str
    job_id: int


class JobStatusResponse(BaseModel):
    """Response with job status"""
    success: bool
    message: Optional[str] = None
    job: Optional[ImageGenerationJobModel] = None
    progress_percentage: Optional[float] = None
    estimated_remaining: Optional[int] = None


class JobLogsResponse(BaseModel):
    """Response with job logs"""
    success: bool
    job_id: int
    logs: List[ImageGenerationLogModel]
    count: int


class JobStatisticsResponse(BaseModel):
    """Response with overall job statistics"""
    success: bool = True
    total_jobs: int
    completed_jobs: int
    running_jobs: int
    failed_jobs: int
    stopped_jobs: int
    total_images_generated: int
    total_images_failed: int
    recipes_without_images: int
    recent_jobs: List[Dict[str, Any]]


class HealthCheckResponse(BaseModel):
    """Health check response"""
    success: bool = True
    message: str
    active_job: Optional[ImageGenerationJobModel] = None
    recipes_without_images: int
    s3_configured: bool
    gemini_configured: bool

File: dashboard_recipe/src/types/imageGeneration.ts
/**
 * Image Generation System Types
 * ==============================
 * TypeScript types for the image generation admin dashboard
 */

export type JobStatus = 'pending' | 'running' | 'completed' | 'stopped' | 'failed';
export type ImageType = 'main' | 'steps' | 'all';
export type LogLevel = 'INFO' | 'WARNING' | 'ERROR' | 'SUCCESS' | 'DEBUG';

export interface ImageGenerationJob {
  id: number;
  status: JobStatus;
  image_type: ImageType;
  started_at: string | null;
  completed_at: string | null;
  total_recipes: number;
  completed_count: number;
  failed_count: number;
  skipped_count: number;
  current_recipe_id: number | null;
  current_recipe_name: string | null;
  last_processed_recipe_id: number | null;
  start_from_recipe_id: number | null;
  should_stop: boolean;
  error_message: string | null;
  error_count: number;
  created_at: string;
  updated_at: string;
}

export interface ImageGenerationLog {
  id: number;
  job_id: number;
  timestamp: string;
  level: LogLevel;
  message: string;
  recipe_id: number | null;
  recipe_name: string | null;
  error_details?: any;
  metadata?: any;
}

export interface JobStatusResponse {
  success: boolean;
  message?: string;
  job: ImageGenerationJob | null;
  progress_percentage?: number;
  estimated_remaining?: number;
}

export interface JobLogsResponse {
  success: boolean;
  job_id: number;
  logs: ImageGenerationLog[];
  count: number;
}

export interface JobStatistics {
  success: boolean;
  total_jobs: number;
  completed_jobs: number;
  running_jobs: number;
  failed_jobs: number;
  stopped_jobs: number;
  total_images_generated: number;
  total_images_failed: number;
  recipes_without_images: number;
  recent_jobs: any[];
}

export interface HealthCheckResponse {
  success: boolean;
  message: string;
  active_job: ImageGenerationJob | null;
  recipes_without_images: number;
  s3_configured: boolean;
  gemini_configured: boolean;
}

export interface StartJobRequest {
  image_type: ImageType;
  start_from_recipe_id?: number;
}

export interface StartJobResponse {
  success: boolean;
  message: string;
  job_id?: number;
  completed?: number;
  failed?: number;
  skipped?: number;
}

export interface StopJobRequest {
  job_id: number;
}

export interface StopJobResponse {
  success: boolean;
  message: string;
  job_id: number;
}

export interface RecipesWithoutImagesResponse {
  success: boolean;
  count: number;
  next_recipe: {
    id: number;
    name: string;
  } | null;
}
